### PR TITLE
Remove async AST node variants for `with`, `for`, and `def`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Per the suggestion in https://github.com/astral-sh/ruff/discussions/6183, this PR removes `AsyncWith`, `AsyncFor`, and `AsyncFunctionDef`, replacing them with an `is_async` field on the non-async variants of those structs. Unlike an interpreter, we _generally_ have identical handling for these nodes, so separating them into distinct variants adds complexity from which we don't really benefit. This can be seen below, where we get to remove a _ton_ of code related to adding generic `Any*` wrappers, and a ton of duplicate branches  for these cases.

## Test Plan

`cargo test` is unchanged, apart from parser snapshots.

### CODE REVIEW

**Comment by charliermarsh** [comment]
You love to see it:

Screen Shot 2023-08-05 at 6 56 58 PM

**Comment by charliermarsh** [comment]
(The formatter stability error is happening on `main`.)

**Review by MichaReiser — APPROVED** [review]
Nice, so much deleted code. 

How did you ensure we use `async: false` in all places that should only apply for sync functions?

**Review by konstin — APPROVED** [review]
:100:

And the changes are much less invasive that i feared

File: crates/ruff/src/checkers/ast/analyze/statement.rs [hunk 1]
```diff
@@ -151,11 +143,11 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
                 flake8_pyi::rules::non_self_return_type(
                     checker,
                     stmt,
+                    *is_async,
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/checkers/ast/analyze/statement.rs:146]: Yes, I will.

File: crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs [hunk 2]
```diff
@@ -217,8 +213,7 @@ impl<'a, 'b> StatementVisitor<'b> for InnerForWithAssignTargetsVisitor<'a, 'b> {
         // Decide whether to recurse.
         match stmt {
             // Don't recurse into blocks that create a new scope.
-            Stmt::ClassDef(_) => {}
-            Stmt::FunctionDef(_) => {}
+            Stmt::ClassDef(_) | Stmt::FunctionDef(_) => {}
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs:216]: No, this was just wrong before. I suspect we had this mistake in a bunch of places.

**Pull request merged**